pkg/controller/deployment/deployment_controller.go

```diff
@@ -22,79 +22,314 @@ import (
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dc.client.ReplicationControllers(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return dc.client.ReplicationControllers(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&api.ReplicationController{},
+		ControllerRelistPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dc.addRC,
+			UpdateFunc: dc.updateRC,
+			DeleteFunc: dc.deleteRC,
+		},
+	)
+
+	// We do not event on anything from the podController, but we use the local
+	// podStore to make queries about the current state of pods (e.g. whether
+	// they are ready or not) more efficient.
+	dc.podStore.Store, dc.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dc.client.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return dc.client.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&api.Pod{},
+		PodRelistPeriod,
+		framework.ResourceEventHandlerFuncs{},
+	)
+
+	dc.syncHandler = dc.syncDeployment
+	dc.podStoreSynced = dc.podController.HasSynced
+	return dc
+}
+
+// Run begins watching and syncing.
+func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
+	defer util.HandleCrash()
+	go dc.dController.Run(stopCh)
+	go dc.rcController.Run(stopCh)
+	go dc.podController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(dc.worker, time.Second, stopCh)
 	}
+	<-stopCh
+	glog.Infof("Shutting down deployment controller")
+	dc.queue.ShutDown()
 }
 
-func (d *DeploymentController) Run(syncPeriod time.Duration) {
-	go util.Until(func() {
-		errs := d.reconcileDeployments()
-		for _, err := range errs {
-			glog.Errorf("Failed to reconcile: %v", err)
-		}
-	}, syncPeriod, util.NeverStop)
+// addRC enqueues the deployment that manages an RC when the RC is created.
+func (dc *DeploymentController) addRC(obj interface{}) {
+	rc := obj.(*api.ReplicationController)
+	if d := dc.getDeploymentForRC(rc); rc != nil {
+		dc.enqueueDeployment(d)
+	}
 }
 
-func (d *DeploymentController) reconcileDeployments() []error {
-	list, err := d.expClient.Deployments(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+// getDeploymentForRC returns the deployment managing the given RC.
+// TODO: Surface that we are ignoring multiple deployments for a given controller.
+func (dc *DeploymentController) getDeploymentForRC(rc *api.ReplicationController) *extensions.Deployment {
+	deployments, err := dc.dStore.GetDeploymentsForRC(rc)
 	if err != nil {
-		return []error{fmt.Errorf("error listing deployments: %v", err)}
+		glog.V(4).Infof("No deployments found for replication controller %v, deployment controller will avoid syncing", rc.Name)
```

Comment:
This error message does not seem right.
We are not checking for `len(deployments) == 0`